Fixes #3579/#3477/#3597 - adjustments to restart logic

kolloch opened the pull request.

aquamatthias commented on src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala:
```diff
@@ -169,9 +163,43 @@ class TaskReplaceActor(
+      if (app.isResident) {
+        // Kill enough tasks so that we end up with end up with one task below minHealthy.
+        // TODO: We need to do this also while restarting, since the kill could get lost.
+        nrToKillImmediately = runningTasksCount - minHealthy + 1
```
`runningTasksCount` could be below `minHealthy`
suggestion: max(runningTasksCount, minHealthy) - minHealthy + 1

aquamatthias commented on src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala:
```diff
@@ -169,9 +163,43 @@ class TaskReplaceActor(
+    var nrToKillImmediately = math.max(0, runningTasksCount - minHealthy)
+
+    if (minHealthy == maxCapacity) {
+      if (app.isResident) {
```
Just to make sure: according to our validation rules, this will never be true, right?

kolloch replied:
No, actually in our standard case that will be true.

minHealthy = (1 * 0.5).ceil.toInt = 1
maxCapacity = (1 * (1 + 0)).toInt = 1

aquamatthias commented on src/test/scala/mesosphere/marathon/upgrade/RestartStrategyTest.scala:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  test("strategy for normal app with 1 instance") {
+    Given("A resident app")
```
which is not resident

aquamatthias commented on src/test/scala/mesosphere/marathon/upgrade/RestartStrategyTest.scala:
```diff
@@ -0,0 +1,132 @@
+  }
+
+  test("maxCapacity strategy for normal app is not exceeded when a task can be killed immediately") {
+    Given("A resident app")
```
s/resident/normal

aquamatthias commented on src/test/scala/mesosphere/marathon/upgrade/RestartStrategyTest.scala:
```diff
@@ -0,0 +1,132 @@
+    Then("the app instance count is exceeded by one")
+    strategy.maxCapacity shouldBe 2
+
+    And("we kill 0 tasks immediately")
```
s/0/1

aquamatthias commented:
@kolloch the integration tests still fail for:
-ResidentTaskIntegrationTest: persistent volume will be re-attached and keep state 
-AppDeployIntegrationTest: backoff delays are reset on configuration changes
-AppDeployIntegrationTest: restarting an app with backoff delay starts immediately
Will rerun to see if this is related to this PR

kolloch commented:
The tests seem to fail quite reproducible -- sorry for not checking after the last change. I will investigate.

aquamatthias commented on src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala:
```diff
@@ -122,7 +122,7 @@ private[launcher] class OfferProcessorImpl(
     def saveTask(taskOpWithSource: TaskOpWithSource): Future[Option[TaskOpWithSource]] = {
       val taskId = taskOpWithSource.taskId
-      log.info("Persisting TaskStateOp for [{}]", taskOpWithSource.taskId)
+      log.info("Processing {} for [{}]", Array[Object](taskOpWithSource.op.stateOp, taskOpWithSource.taskId): _*)
```
Nit: we are using log with format and `s""` randomly. Seeing this call I would opt for `s""`

kolloch replied:
ok

aquamatthias commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -278,10 +278,11 @@ object Task {
   def tasksById(tasks: Iterable[Task]): Map[Task.Id, Task] = tasks.iterator.map(task => task.taskId -> task).toMap
 
-  case class Id(idString: String) {
+  case class Id(idString: String) extends Comparable[Id] {
```
Nit: `Ordered[Id]` would also give you equality operators.

aquamatthias replied:
BTW: why we need ordering for Id?

kolloch replied:
Nice, I didn't really think of `Ordered`.

I think that it is always nice to be able to sort things, e.g. for reproducibility.

aquamatthias commented on src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala:
```diff
@@ -169,9 +163,50 @@ class TaskReplaceActor(
+  private[upgrade] def computeRestartStrategy(app: AppDefinition, runningTasksCount: Int): RestartStrategy = {
+    // in addition to an app definition which passed validation, we require:
+    require(app.instances > 0, s"instances must be > 0 but is ${app.instances}")
+    require(runningTasksCount >= 0, s"running task count must be >=0 but is $runningTasksCount")
```
Not sure if this requirement is valid:
- deploy application with n tasks
- the deployment is successful
- all tasks fail, all restart attempts fail as well
- we deploy a new version at exactly a time all tasks are gone

aquamatthias commented on src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala:
```diff
@@ -169,9 +163,50 @@ class TaskReplaceActor(
+      s"$minHealthy tasks running, maximum capacity $maxCapacity, killing $nrToKillImmediately of " +
+      s"$runningTasksCount running tasks immediately")
+
+    assume(nrToKillImmediately >= 0, s"nrToKillImmediately must be >=0 but is $nrToKillImmediately")
```
If the requirement above is not true, than also this assume would not hold.

kolloch replied:
No, we cannot. It also depends on the number of running tasks.

E.g. if we have 3 running tasks, and a min health capacity of 4 and a max capacity of 4, we still would not need to kill a task.

kolloch replied:
We could assume something like `minHealthy < maxCapacity || runningTasksCount - nrToKillImmediately < maxCapacity`

aquamatthias commented:
@kolloch Looks very good to me - especially the green box above. Thanks!
Please squash your commits and I will merge.

kolloch commented on src/main/scala/mesosphere/marathon/upgrade/TaskReplaceActor.scala:
```diff
@@ -174,23 +174,38 @@ object TaskReplaceActor {
+    assume(nrToKillImmediately >= 0, s"nrToKillImmediately must be >=0 but is $nrToKillImmediately")
+    assume(maxCapacity > 0, s"maxCapacity must be >0 but is $maxCapacity")
+    def canStartNewTasks: Boolean = minHealthy < maxCapacity || runningTasksCount - nrToKillImmediately < maxCapacity
+    assume(canStartNewTasks, s"must be able to start new tasks")
```
@aquamatthias does that look good to you?

kolloch commented:
I split up and squashed the commits along authors and topics.

The integration tests fail on teamcity because they cannot resolve a necessary GPG key. On my local machine, I could get beyond that step. Not sure how to proceed:

* Hope that it is a temporary TeamCity network-related problem and wait?
* Investigate, trying alternatives.

It is annoying to debug if we cannot reproduce it locally. We had similar problems already in the past.

kolloch commented:
aah, the last teamcity build moved beyond that error.

kolloch commented:
Ok, there was another sporadic error which looks similar to the ones that I tried to fix.

I think that the code in this PR is an improvement anyway. I will try to reproduce that error locally with extended debug settings, again.

In the mean time I would suggest to

* either merge this PR completely
* or to separate the last commit that tries to address the error seen in TeamCity but which apparently is not complete and merge the rest

aquamatthias commented:
Thanks!
